### ISSUE TITLE
How to understand conversion:current

### ISSUE THREAD

**Issue by dvorak0** [beancount/fava#970]
I see the figure 1 when I open `http://localhost:5000/beancount/balance_sheet/?conversion=CNY&time=2017`.
![Capture](https://user-images.githubusercontent.com/2220369/65390487-1479c880-dd92-11e9-8067-18c99fe136a8.PNG)

Where can I find docs about `Equity:Conversion:Previous`? I expect it to be 'conversion happened previously'. I guess the `Equity:Opening-Balances` is implemented as 'conversion happened previously'.

Thanks for any help. Attached is the bean file if it's helpful.

```
option "operating_currency" "CNY"
plugin "beancount.plugins.implicit_prices"


2015-01-01 open Assets:Stock
2015-01-01 open Assets:Cash
2015-01-01 open Income:Salary


2015-04-01 * "income"
  Assets:Cash +10000.00 CNY
  Income:Salary -12000.00 HKD @@ 10000.00 CNY
  Assets:Cash +2000.00 HKD
  Income:Salary -2000.00 HKD

2016-04-01 * "income"
  Assets:Cash +11000.00 CNY
  Income:Salary -12000.00 HKD @@ 11000.00 CNY
  Assets:Cash +2000.00 HKD
  Income:Salary -2000.00 HKD

2017-04-01 * "income"
  Assets:Cash +12000.00 CNY
  Income:Salary -12000.00 HKD @@ 12000.00 CNY
  Assets:Cash +2000.00 HKD
  Income:Salary -2000.00 HKD
  

2015-05-01 * "buy"
  Assets:Cash
  Assets:Stock +100.00 PINGAN {50.00 CNY}
  
2016-05-01 * "buy"
  Assets:Cash
  Assets:Stock +100.00 PINGAN {60.00 CNY}
  
2017-05-01 * "buy"
  Assets:Cash
  Assets:Stock +100.00 PINGAN {70.00 CNY}
  
  
2018-05-01 price PINGAN              100 CNY

2018-05-01 price HKD              1.2 CNY
```

**Comment by yagebu** [comment]
I always think of all the Equity subaccounts as being there to ensure that everything sums up to 0.

- Opening-Balances: You probably start your Beancount file with some money in your bank account already (or some sort of asset) - that's balanced out by your opening balances.
- Conversions: If you buy something at cost (with `{}`) and sell it a different cost, that difference is reflected in the conversions equity subaccount.
- Earnings: basically just Income + Expenses

The current/previous is - as you guessed - the split between the entries in 2017 and those before.

If you want to see how this is implemented, you can check out the summarize module in Beancount and the core.tree module in Fava.

**Comment by dvorak0** [comment]
@yagebu I agree with you on the principles but am still conflused with details.

As you can see in figure 1, `conversion:previous` is empty which doesn't follow your 

> The current/previous is - as you guessed - the split between the entries in 2017 and those before.

In fact, the `opening-balance`(3000 CNY) should be counted as `conversion:previous`. The current implementation is confusing in mixed `opening-balance` with `conversion:previous`

**Comment by yagebu** [comment]
I rarely look at the Equity accounts, so I also have to look up how they work exactly. When you set a time filter, we use the clamp function from beancount.core.summarize. It transfers all income and expenses to the earnings account, and summarizes all transactions against the opening balances account (so that the balance sheet accounts start at their correct balances). The conversions are inserted at the end of the time interval. You can see the generated entries for the time filter by looking at the Journal and showing all transactions (toggling the `x`).

> As you can see in figure 1, conversion:previous is empty which doesn't follow your

This happens because you've selected "Converted to CNY" - the conversion entries that Beancount inserts for the previous conversions have a price of `0 NOTHING`, so they get converted to 0.

**Comment by dvorak0** [comment]
@yagebu `Conversion:previous` should not be 0. I have two conversions in year 2015(`-12000.00 HKD @@ 10000.00 CNY`) and 2016(`-12000.00 HKD @@ 11000.00 CNY`). That leads -3000 CNY when price of HKD is 1.00 CNY in year 2017. Then `Conversion:previous` should be -3000 CNY, which however is `opening-balance` in current implementation.

**Comment by yagebu** [comment]
Ah, it seems that Beancount inserts the conversion entry for all conversions happening in 2017 in your case (which gets correctly converted to 0) - I find it sensible that it only considers conversions in the current period but odd that it uses the "previous" and not the "current" account.

@blais: When a time filter is set in Fava, we use beancount.core.summarize.clamp_opt to obtain the list of entries used in all reports. It does the following (let's say for a time filter of 2017):

- Transfers all income and expenses prior to 2017 to Equity:Earnings:Previous.
- Summarizes all balances prior to 2017, against Equity:Opening-Balances
- Removes all entries after 2017
- Inserts a conversion entry for all conversion occuring in 2017, against the account Equity:Conversions:Previous

The last step I find a bit odd, shouldn't it use Equity:Conversions:Current?

(for the balance sheet, Fava inserts some more account balances using the "current" accounts and the unrealized gains account, see https://github.com/beancount/fava/blob/ee455e7644765cffd18d89e3bfbb18a866cbdf50/fava/core/tree.py#L134)

**Comment by yagebu** [comment]
The inconsistent naming happens in Beancount, so I'll close this issue

**Comment by Smart-Hypercube** [comment]
Hi @dvorak0 ,

Just want to let you know, I found this issue and [your mail thread with Blais](https://www.mail-archive.com/beancount@googlegroups.com/msg02617.html) by searching this problem online. I too have been troubled by this problem for a long time. As I increasingly believe this is indeed a bug, I opened an issue https://github.com/beancount/beancount/issues/494 today and got this bug fixed~ I still wonder why no one has noticed their balance sheets have such a big problem and wanted to fix it o_0

I have confirmed that once you install the fixed version of beancount, balance sheets generated by fava are also fixed. So no need to change fava's code.
